File: app/src/main/java/io/github/fvasco/pinpoi/util/PlacemarkDistanceComparator.java
package io.github.fvasco.pinpoi.util;

import android.location.Location;

import java.util.Comparator;
import java.util.Objects;

import io.github.fvasco.pinpoi.model.Placemark;

/**
 * Compare placemark using distance from a specific {@linkplain android.location.Location}
 *
 * @author Francesco Vasco
 */
public class PlacemarkDistanceComparator implements Comparator<Placemark> {

    private final Location center;
    private final float[] distanceResult = new float[1];

    public PlacemarkDistanceComparator(final Location center) {
        Objects.requireNonNull(center);
        this.center = center;
    }

    @Override
    public int compare(Placemark lhs, Placemark rhs) {
        int res = Double.compare(calculateDistance(lhs), calculateDistance(rhs));
        if (res == 0)
            res = Long.compare(lhs.getId(), rhs.getId());
        return res;
    }

    /**
     * Calculate distance to placemark
     * {@see Location@distanceTo}
     */
    public double calculateDistance(final Placemark p) {
        Location.distanceBetween(center.getLatitude(), center.getLongitude(), p.getLatitude(), p.getLongitude(), distanceResult);
        return distanceResult[0];
    }

}

File: app/src/main/java/io/github/fvasco/pinpoi/util/PlacemarkDistanceComparator.kt
package io.github.fvasco.pinpoi.util

import io.github.fvasco.pinpoi.model.PlacemarkSearchResult

/**
 * Compare placemark using distance from a specific [Coordinates]

 * @author Francesco Vasco
 */
class PlacemarkDistanceComparator(private val center: Coordinates) :
    Comparator<PlacemarkSearchResult> {

    override fun compare(plhs: PlacemarkSearchResult, prhs: PlacemarkSearchResult): Int {
        val lhs = plhs.coordinates
        val rhs = prhs.coordinates
        var res = center.distanceTo(lhs).compareTo(center.distanceTo(rhs))
        if (res == 0) {
            // equals <==> same coordinates
            res = lhs.latitude.compareTo(rhs.latitude)
            if (res == 0) res = lhs.longitude.compareTo(rhs.longitude)
        }
        return res
    }
}
